crates/python/src/qpu/api.rs

```diff
@@ -187,17 +187,82 @@ py_function_sync_async! {
         job_id: String,
         quantum_processor_id: String,
         client: Option<PyQcsClient>,
-        endpoint_id: Option<String>,
+        execution_options: Option<PyExecutionOptions>
     ) -> PyResult<ExecutionResults> {
         let client = PyQcsClient::get_or_create_client(client).await;
 
-        let job_target = endpoint_id.map_or_else(|| JobTarget::QuantumProcessorId(quantum_processor_id), JobTarget::EndpointId);
-
-        let results = qcs::qpu::api::retrieve_results(job_id.into(), &job_target, &client)
+        let results = qcs::qpu::api::retrieve_results(job_id.into(), quantum_processor_id, &client, execution_options.unwrap_or_default().as_inner())
             .await
             .map_err(RustRetrieveResultsError::from)
             .map_err(RustRetrieveResultsError::to_py_err)?;
 
         Ok(results.into())
     }
 }
+
+py_wrap_type! {
+    #[derive(Debug, Default)]
+    PyExecutionOptions(ExecutionOptions) as "ExecutionOptions"
+}
+impl_repr!(PyExecutionOptions);
+impl_as_mut_for_wrapper!(PyExecutionOptions);
+
+#[pymethods]
+impl PyExecutionOptions {
+    #[staticmethod]
+    fn default() -> Self {
+        Self::from(ExecutionOptions::default())
+    }
+}
+
+py_wrap_type! {
+    PyExecutionOptionsBuilder(ExecutionOptionsBuilder) as "ExecutionOptionsBuilder"
+}
+
+#[pymethods]
+impl PyExecutionOptionsBuilder {
+    #[new]
+    fn new() -> Self {
+        Self::default()
+    }
+
+    #[staticmethod]
+    fn default() -> Self {
+        Self::from(ExecutionOptionsBuilder::default())
+    }
+
+    fn connection_strategy(&self, connection_strategy: PyConnectionStrategy) -> Self {
+        // `derive_builder::Builder` doesn't implement AsMut, meaning we can't use `PyWrapperMut`,
+        // which forces us into this awkward clone.
+        Self::from(
+            self.as_inner()
+                .clone()
+                .connection_strategy(connection_strategy.as_inner().clone())
+                .clone(),
+        )
+    }
```

Comment:
I don't think there is any way around doing at least two clones here. We can't return `&Self` because of Python/pyo3, and there is no way to get the inner `ConnectionStrategy` directly.

Reply:
What if the python-side builder were just more limited and didn't allow chaining? That is, `fn set_connection_strategy(...) -> () `?

Reply:
I ended up going with this suggestion because I realized, as written, it had a sharp edge. If you weren't using chaining, the inner builder wasn't actually updated, so if you instead opted to line by line update the builder instead of chaining, and threw away the returned builders along the way, you would effectively be doing nothing.

Restricting the return type lets us implement it as a setter, which forces users into setting it as a property and gives user one clear, correct way of doing it. Unfortunately, it doesn't reduce the amount of clones needed